### PR TITLE
Fix #9792: PD tasks fail to launch if body is inside Part and activated

### LINKED ISSUE #9792: [Problem] PartDesign tasks fail to launch when body is inside Part and activated
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

0.21 (Development)

### Full version info

```shell
OS: Debian GNU/Linux 11 (bullseye) (X-Cinnamon/lightdm-xsession)
Word size of FreeCAD: 64-bit
Version: 0.21.0.33411 (Git)
Build type: Release
Branch: master
Hash: 127404d793d5c2081383b46a7b677426d9a0c5aa
Python 3.9.2, Qt 5.15.2, Coin 4.0.0, Vtk 9.0.1, OCC 7.7.1
Locale: English/United States (en_US)
Installed mods:
```

### Subproject(s) affected?

PartDesign

### Problem description

-new document
-create Part
-create Body, body is automatically placed inside Part however it is not highlighted as activated
(if you try now to create a sketch in this moment it works without problems)
-double click on Body to activate it
-attempt creating a sketch on body (or launch any PD task really), the task is not launched and the error shown in report view

> 14:57:31  <Gui> Document.cpp(385): Invalid sub object 'Unnamed#Part.Body.Box.Body.'

double clicking on the created sketch or feature successfully launchs the task.

### Anything else?

originally reported at https://forum.freecad.org/viewtopic.php?p=688246#p688246

and here the example file provided there (remove .zip extension):
[Yoke.FCStd.zip](https://github.com/FreeCAD/FreeCAD/files/11782390/Yoke.FCStd.zip)

a3c4f4bb01f5201daf666852444a8f9c5adbda8d is mentioned as introducing this problem. @wwmayer got any ideas?

If we can't fix this before release this commit should be reverted.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by adrianinsaval**
introduced by PR #9538

Thank you for creating a pull request to contribute to FreeCAD! Place an "X" in between the brackets below to "check off" to confirm that you have satisfied the requirement, or ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10) if there is something you don't understand.

- [x]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

### CODE REVIEW

**Comment by adrianinsaval** [comment]
this fixes a critical bug that is blocking release, @0penBrain if you have time could you review?

**Comment by grubertechnic** [comment]
A test here works flawless.

```
OS: Debian GNU/Linux 11 (bullseye) (X-Cinnamon/lightdm-xsession)
Word size of FreeCAD: 64-bit
Version: 0.21.0.33487 (Git)
Build type: Release
Branch: PD9792
Hash: ab243719252bd6af6fdbeec15fa82bb94415a601
Python 3.9.2, Qt 5.15.2, Coin 4.0.0, Vtk 9.0.1, OCC 7.7.1
Locale: English/United States (en_US)
Installed mods: 
  * QuickMeasure 2022.10.28
  * Assembly4 0.12.6
  * CurvedShapes 1.0.4
  * Curves 0.6.8
```

**Comment by 0penBrain** [comment]
Tested but I can't see any difference between current and PR code. :confused: 
Actually I was not able to get any problem with current code, trying to understand what this PR changes.
Could one help by providing a way to reproduce the current issue that is solved here ?

**Comment by adrianinsaval** [comment]
step by step instructions are at #9792

**Comment by 0penBrain** [comment]
> step by step instructions are at #9792

OK, I tested.

And yes, the PR fixes the error message (hence probably reach its target).

But well, fundamentally I don't think #9538 creates this bug, but actually revealed a deeper underlying one.

Look at this.

 * New document
 * Create a Part container -> OK, container is active
 * Create a Body -> Hmmm, it's created inside the Part but in a very weird state. It is displayed bold, but do not have the "active" background color :confused:

Still more strange
 * New document
 * Create a Part container
 * Create a Body
 * Double click the Body (take care to not deselect it) -> It goes to **inactive** state

compared to
 * New document
 * Create a Part container
 * Create a Body
 * Deselect the Body by clicking empty area of the tree then double-click on it -> It goes to **active** state

At this time I really feel like when a Body is created when a Part container is active, it is created in a very weird state.
I think this is what we have to solve now. Feeling like this PR is patching the consequence of the bug, not the bug itself.

**Comment by adrianinsaval** [comment]
No I think there are two bugs at play here IMO the one fixed here is also a bug because it happens when the body is in the correct state of being active.

You mention a bug where bodies are created in a weird state and I agree but I don't think that's the root of the problem for the reported bug, but still should be fixed.

**Comment by adrianinsaval** [comment]
https://forum.freecad.org/viewtopic.php?p=691905#p691905
werner says:
> What happens is that

> ```
> auto activeBody = activeView->getActiveObject<PartDesign::Body*>(PDBODYKEY,&parent,&subname);
> ```
> unexpectedly sets subname to "Body". Afterwards when calling getParent() then the string "Body.Sketch." is prepended and "." appended which results into the incorrect subname "Body.Sketch001.Body.".
> 
> Fixing the issues is easy: before calling getParent() the string subname must be cleared.

**Comment by 0penBrain** [comment]
> https://forum.freecad.org/viewtopic.php?p=691905#p691905 werner says:

Fully agree with him. I pushed to your branch a commit that implements this. This looks IMO simpler and cleaner.
Feel free to squash if OK.

What was the reason for you to implement something else ?

**Comment by adrianinsaval** [comment]
These functions are very confusing to me, why is it unexpected that `subname` is set to Body, that seems like the logical outcome of line 92 to me, if we don't want this to modify `subname` why are we passing it as argument at all? Following that same line of thought, if we are going to get `parent` independently anyways why are we passing it as an argument to `getActiveObject` anyway? Perhaps in that line we should just let it use their default nullptr value and just always get parent and subname later.

The behavior of `getActiveObject` also seems strange to me, when you initially create the Part and the Body, parent=Body and subname is empty, but if you then double click on the Body parent=Part and subname=Body, this later behavior seems the most correct to me, but it's not handled correctly by setEdit. Maybe when you first create the Body something goes wrong and it doesn't properly register that it has a Part as parent, the fact the Body is not highlighted makes it suspicious too.

And why does a function called `getParent` modify `subname`?

File: src/Mod/PartDesign/Gui/Utils.cpp [hunk 1]
```diff
@@ -90,9 +90,9 @@ bool setEdit(App::DocumentObject *obj, PartDesign::Body *body) {
     App::DocumentObject *parent = nullptr;
     std::string subname;
     auto activeBody = activeView->getActiveObject<PartDesign::Body*>(PDBODYKEY,&parent,&subname);
+    subname.clear();
     if (activeBody != body) {
```
Comment (adrianinsaval) [inline src/Mod/PartDesign/Gui/Utils.cpp:94]: I'm also not understanding in which situation this statement is true. Anybody knows?

**Comment by adrianinsaval** [comment]
This seemed to work and makes a little more sense to me, what do you think?

File: src/Mod/PartDesign/Gui/Utils.cpp (hunk 1 above)
Comment (0penBrain) [reply src/Mod/PartDesign/Gui/Utils.cpp:93]: I don't know too.

**Comment by 0penBrain** [comment]
> These functions are very confusing to me, why is it unexpected that `subname` is set to Body, that seems like the logical outcome of line 92 to me, if we don't want this to modify `subname` why are we passing it as argument at all? Following that same line of thought, if we are going to get `parent` independently anyways why are we passing it as an argument to `getActiveObject` anyway? Perhaps in that line we should just let it use their default nullptr value and just always get parent and subname later.

Agree. If codebase is correct, it should be possible (and better) to just call `auto activeBody = activeView->getActiveObject<PartDesign::Body*>(PDBODYKEY)` Would you mind testing ?

> The behavior of `getActiveObject` also seems strange to me, when you initially create the Part and the Body, parent=Body and subname is empty, but if you then double click on the Body parent=Part and subname=Body, this later behavior seems the most correct to me, but it's not handled correctly by setEdit. Maybe when you first create the Body something goes wrong and it doesn't properly register that it has a Part as parent, the fact the Body is not highlighted makes it suspicious too.
> 
> And why does a function called `getParent` modify `subname`? Given that this function is actually setting two variables, won't it be better to have it be something like:
> 
> ```c++
> getParentAndSubname(obj, &parent, &subname);
> ```
> 
> to be more explicit about what it does

Yes, not everything is ideally implemented, but here IMO it's too much for this PR to go fixing all this. :)

**Comment by adrianinsaval** [comment]
> Agree. If codebase is correct, it should be possible (and better) to just call `auto activeBody = activeView->getActiveObject<PartDesign::Body*>(PDBODYKEY)` Would you mind testing ?

I did! It's working. The PR is that now.

> Yes, not everything is ideally implemented, but here IMO it's too much for this PR to go fixing all this. :)

yep, PR remains limited to the fix

**Pull request merged**